Declarative contribution of custom Tree Explorer

octref opened the pull request:
PR for #12163.
### Main changes:
- `contributes.explorer` section for letting extension add custom tree explorers
- `TreeExplorerNodeProvider<T>` in vscode namespace
- `registerTreeExplorerNodeProvider(providerId: string, provider: TreeExplorerNodeProvider<any>): Disposable` in vscode namespace
- `vs/workbench/parts/explorers/*`
  - The viewlet for tree explorers
  - Command to toggle each tree explorer
### Not yet implemented:
- [x] Handle long running `TreeExplorerNodeProvider.provideRootNode` and `TreeExplorerNodeProvider.resolveChildren` by adding a progress bar.
- [x] Handle restoring external viewlet if last active viewlet is external.
  - Construct blank sidebar and restore external viewlet after it gets registered
  - Show progress bar at top of the blank sidebar while waiting for external viewlet getting registered
- [x] Refresh button at top of external viewlet
- [x] Show error popup for handling errors in `TreeExplorerNodeProvider`'s methods.
- [x] Show error popup for no matching `TreeExplorerNodeProvider` found
- [x] Use `-webkit-filter` and opacity to style user defined activity bar icons to indicate activation.
- [ ] Add extension that test against the API
### High level flow
- Workbench get constructed
- Stock Viewlets get registered
- Extensions' `package.json` get read.
- `ViewletDescriptor` for external viewlets get registered at `ViewletRegistry`
- `ActivitybarPart` adds icons for external viewlets when external viewlets finish registration 
- When icon for external viewlet get clicked, construct the viewlet and hook it to the registered provider
### Examples
- [Deps](https://github.com/octref/explorer-api-example/tree/master/dep): easiest to get running.
- [Avatar](https://github.com/octref/explorer-api-example/tree/master/avatar): need to get a GH token. See `src/token.ts`.
- [Redis](https://github.com/octref/explorer-api-example/tree/master/redis): need to install redis.

joaomoreno commented:
Forwarding to @bpasero since he owns the Explorer and to @Tyriar since he is a buddy.

bpasero commented:
@octref can you please resolve the merge conflicts?

octref commented:
@bpasero Rebased on master.

Tyriar commented on src/vs/vscode.d.ts:
```diff
@@ -1349,6 +1349,72 @@ declare module 'vscode' {
+		/**
+		 * Get the command to execute when `node` is clicked.
+		 *
+		 * Commands can be registered through (#commands.registerCommand). `node` will be provided
```
`[registerCommand](#commands.registerCommand)`

Tyriar commented on src/vs/vscode.d.ts:
```diff
@@ -1349,6 +1349,72 @@ declare module 'vscode' {
+		/**
+		 * Determine if `node` has children and is expandable.
+		 *
+		 * Default to return `true` if not provided.
```
> Defaults to `true`.

I'd attach this sentence to the first line.

Tyriar commented on src/vs/vscode.d.ts:
```diff
@@ -1349,6 +1349,72 @@ declare module 'vscode' {
+		/**
+		 * Provide a human-readable string that will be used for rendering the node.
+		 *
+		 * Default to use `node.toString()` if not provided.
```
> Defaults to `node.toString()`.

I'd attach this sentence to the first line.

Tyriar commented on src/vs/vscode.d.ts:
```diff
@@ -3804,6 +3870,15 @@ declare module 'vscode' {
 		export function registerTextDocumentContentProvider(scheme: string, provider: TextDocumentContentProvider): Disposable;
 
 		/**
+		 * Register a [tree explorer node provider](#TreeExplorerNodeProvider).
```
tree explorer node provider -> TreeExplorerNodeProvider

Tyriar commented on src/vs/workbench/api/node/extHostTreeExplorers.ts:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	private showErrorMessage(message: string): void {
+		this._proxy.$showMessage(Severity.Error, message);
```
You could probably make ExtHostMessageService a dependency and call into that from here?

Tyriar commented on src/vs/workbench/parts/explorers/common/treeExplorerViewModel.ts:
```diff
@@ -0,0 +1,44 @@
+import { TPromise } from 'vs/base/common/winjs.base';
+import { TreeExplorerNodeProvider } from 'vscode';
+
+export class InternalTreeExplorerNode implements TreeExplorerNodeContent {
```
Why not just TreeExplorerNode?

Tyriar commented on src/vs/workbench/parts/explorers/common/treeExplorerViewModel.ts:
```diff
@@ -0,0 +1,44 @@
+	}
+}
+
+export interface InternalTreeExplorerNodeProvider {
```
Why not just TreeExplorerNodeProvider?

Tyriar commented on src/vs/workbench/api/node/extHostTreeExplorers.ts:
```diff
@@ -0,0 +1,101 @@
+		this._externalNodeMaps = Object.create(null);
+	}
+
+	registerTreeExplorerNodeProvider(providerId: string, provider: TreeExplorerNodeProvider<any>): Disposable {
```
Shouldn't this return `vscode.Disposable` from `import * as vscode from 'vscode';`?

Tyriar commented on src/vs/workbench/api/node/extHostTreeExplorers.ts:
```diff
@@ -0,0 +1,101 @@
+		this._treeExplorerNodeProviders[providerId] = provider;
+
+		return new Disposable(() => {
+			delete this._treeExplorerNodeProviders[providerId];
```
Don't you need to clean up `_externalNodeMaps` too?

Tyriar commented on src/vs/workbench/browser/parts/activitybar/activitybarPart.ts:
```diff
@@ -65,6 +97,30 @@ export class ActivitybarPart extends Part implements IActivityService {
 		}
 	}
 
+	getIsEnabledForRegisteredViewlets(): { [viewletId: string]: boolean } {
```
Add access modifier

Tyriar commented on src/vs/workbench/browser/parts/activitybar/activitybarPart.ts:
```diff
@@ -65,6 +97,30 @@ export class ActivitybarPart extends Part implements IActivityService {
+		return result;
+	}
+
+	toggleViewlet(viewletId: string): void {
```
Add access modifier

Tyriar commented on src/vs/workbench/browser/parts/activitybar/activitybarPart.ts:
```diff
@@ -84,43 +140,77 @@ export class ActivitybarPart extends Part implements IActivityService {
 
 		// Top Actionbar with action items for each viewlet action
-		this.createViewletSwitcher($result.clone());
+		this.createViewletSwitcher($result.clone().addClass('position-top'));
```
What does this do?

Tyriar commented on src/vs/workbench/browser/parts/activitybar/activitybarPart.ts:
```diff
@@ -84,43 +140,77 @@ export class ActivitybarPart extends Part implements IActivityService {
 
 	private createViewletSwitcher(div: Builder): void {
-
-		// Composite switcher is on top
```
Why remove?

Tyriar commented on src/vs/workbench/browser/parts/activitybar/activitybarPart.ts:
```diff
@@ -144,8 +234,12 @@ class ViewletActivityAction extends ActivityAction {
 	private lastRun: number = 0;
 
+	private _onOpenExternalViewlet = new Emitter<string>();
+	get onOpenExternalViewlet(): Event<string> { return this._onOpenExternalViewlet.event; };
```
Access modifier

Tyriar commented on src/vs/workbench/parts/explorers/browser/treeExplorerActions.contribution.ts:
```diff
@@ -0,0 +1,60 @@
+		for (let viewletId in viewletsToggleStataus) {
+			picks.push({
+				id: viewletId,
+				label: (viewletsToggleStataus[viewletId] ? 'Disable ' : 'Enable ') + this.getShortViewletId(viewletId)
```
Is this `viewletId` the `treeExplorerNodeProviderId` in the contribution? If so why aren't you using treeLabel as the label in the quick open?

Tyriar commented on src/vs/workbench/parts/explorers/browser/treeExplorerActions.contribution.ts:
```diff
@@ -0,0 +1,60 @@
+	}
+
+	run(): TPromise<any> {
+		const viewletsToggleStataus = this.activityService.getIsEnabledForRegisteredViewlets();
```
Typo: Stataus

Tyriar commented on src/vs/workbench/parts/explorers/browser/treeExplorerViewlet.ts:
```diff
@@ -0,0 +1,88 @@
+	create(parent: Builder): TPromise<void> {
+		super.create(parent);
+
+		this.viewletContainer = parent.div().addClass('custom-tree-explorer-viewlet');
```
I don't think this class is used.

Tyriar commented on src/vs/workbench/parts/explorers/browser/treeExplorerViewletService.ts:
```diff
@@ -0,0 +1,76 @@
+	executeCommand(providerId: string, node: InternalTreeExplorerNode): TPromise<void>;
+}
+
+export class TreeExplorerViewletService implements ITreeExplorerViewletService {
```
Fill in access modifiers

Tyriar commented on src/vs/workbench/parts/explorers/browser/treeExplorerViewlet.ts:
```diff
@@ -0,0 +1,88 @@
+import { IActivityService } from 'vs/workbench/services/activity/common/activityService';
+import { VIEWLET_ID_ROOT } from 'vs/workbench/parts/explorers/common/treeExplorer';
+
+export class TreeExplorerViewlet extends Viewlet {
```
Fill in access modifiers

Tyriar commented on src/vs/workbench/parts/explorers/browser/views/treeExplorerView.ts:
```diff
@@ -0,0 +1,115 @@
+			return this.treeExplorerViewletService.provideRootNode(this.treeNodeProviderId).then(tree => {
+				this.tree.setInput(tree);
+			});
+		} else {
```
else/indent is redundant

Tyriar commented on src/vs/workbench/parts/explorers/browser/views/treeExplorerViewer.ts:
```diff
@@ -0,0 +1,130 @@
+import { ITreeExplorerViewletService } from 'vs/workbench/parts/explorers/browser/treeExplorerViewletService';
+import { IProgressService } from 'vs/platform/progress/common/progress';
+
+export class TreeDataSource implements IDataSource {
```
Fill in access modifiers

Tyriar commented on src/vs/workbench/parts/explorers/browser/views/treeExplorerViewer.ts:
```diff
@@ -0,0 +1,130 @@
+	getChildren(tree: ITree, node: InternalTreeExplorerNode): TPromise<InternalTreeExplorerNode[]> {
+		const promise = this.treeExplorerViewletService.resolveChildren(this.treeNodeProviderId, node);
+
+		this.progressService.showWhile(promise, 800);
```
Why 800?

Tyriar commented on src/vs/workbench/parts/explorers/browser/views/treeExplorerViewer.ts:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	getContentHeight(tree: ITree, element: any): number {
+		return 22;
```
Why 22?

Tyriar commented on src/vs/workbench/parts/explorers/common/treeExplorerViewModel.ts:
```diff
@@ -0,0 +1,44 @@
+
+	id: number;
+
+	label: string = 'label';
```
Why does this default to `'label'`?

Tyriar commented on src/vs/workbench/parts/explorers/common/treeExplorerViewModel.ts:
```diff
@@ -0,0 +1,44 @@
+import { TreeExplorerNodeProvider } from 'vscode';
+
+export class InternalTreeExplorerNode implements TreeExplorerNodeContent {
+	static idCounter = 1;
```
Should not be public

Tyriar commented on src/vs/workbench/parts/explorers/media/Refresh.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16"><path d="M13.451 5.609l-.579-.939-1.068.812-.076.094c-.335.415-.927 1.341-1.124 2.876l-.021.165.033.163.071.345c0 1.654-1.346 3-3 3-.795 0-1.545-.311-2.107-.868-.563-.567-.873-1.317-.873-2.111 0-1.431 1.007-2.632 2.351-2.929v2.926s2.528-2.087 2.984-2.461h.012l3.061-2.582-4.919-4.1h-1.137v2.404c-3.429.318-6.121 3.211-6.121 6.721 0 1.80 ...
```
Are you reusing an icon here?

Tyriar replied:
I believe we're trying to dedupe them and now a bunch of warnings pop up during the build when inlining duplicate icons.

octref replied:
OK, but I'd rather do it after we put Refresh and the inverse version into a single place. Taking the svg file from `parts/files` into `parts/explorers` don't feel good.

Tyriar commented on src/vs/workbench/parts/explorers/media/Refresh_inverse.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16"><path d="M13.451 5.609l-.579-.939-1.068.812-.076.094c-.335.415-.927 1.341-1.124 2.876l-.021.165.033.163.071.345c0 1.654-1.346 3-3 3-.795 0-1.545-.311-2.107-.868-.563-.567-.873-1.317-.873-2.111 0-1.431 1.007-2.632 2.351-2.929v2.926s2.528-2.087 2.984-2.461h.012l3.061-2.582-4.919-4.1h-1.137v2.404c-3.429.318-6.121 3.211-6.121 6.721 0 1.80 ...
```
Are you reusing an icon here?

octref commented on src/vs/platform/explorers/browser/explorerExtensionPoint.ts:
```diff
@@ -0,0 +1,71 @@
+
+namespace schema {
+
+	export interface IExplorer {
```
Good point.

octref commented on src/vs/platform/explorers/browser/explorerExtensionPoint.ts:
```diff
@@ -0,0 +1,71 @@
+
+		const getIconRule = (iconPath) => { return `background-image: url('${iconPath}')`; };
+		if (icon) {
+			const iconClass = `.monaco-workbench > .activitybar .monaco-action-bar .action-label.${treeExplorerNodeProviderId}`;
```
Nope, activity bar doesn't change with themes for now (I'd like it to though)

octref commented on src/vs/platform/explorers/browser/explorerExtensionPoint.ts:
```diff
@@ -0,0 +1,71 @@
+			const iconPath = join(extension.description.extensionFolderPath, icon);
+			createCSSRule(iconClass, getIconRule(iconPath));
+			// Coerce the icon into a style similar to stock icons
+			createCSSRule(iconClass, '-webkit-filter: grayscale(1) invert(1)');
```
![image](https://cloud.githubusercontent.com/assets/4033249/19657526/7176fdfa-99d9-11e6-9341-a72458b887ab.png)

Tyriar replied:
I mean pull it into the CSS file, not done in typescript since it's static.

octref commented on src/vs/vscode.d.ts:
```diff
@@ -1349,6 +1349,72 @@ declare module 'vscode' {
+		 *
+		 * @return The root node.
+		 */
+		provideRootNode(): T | Thenable<T>;
```
I thought about it, the thing is the invisible root node actually holds the label for here:

`vscode` and `open editors`

![image](https://cloud.githubusercontent.com/assets/4033249/19657815/7ab6280e-99da-11e6-933c-de8d5317a9d0.png)

I feel later we are gonna consider letting extension contribute multi section viewlets, and I want to leave it here so they can specify label for each section later.

octref commented on src/vs/vscode.d.ts:
```diff
@@ -1349,6 +1349,72 @@ declare module 'vscode' {
+	 * - hasChildren: Whether the node has children and is expandable.
+	 * - clickCommand: A command to execute when the node is clicked.
+	 */
+	export interface TreeExplorerNodeProvider<T> {
```
Thanks for the feedback!

---

> Not a big friend of the lonely \<T\>

`TreeItem` was what I was initially using. However after talking to Kai and Daniel we feel the current approach is cleaner. The main advantages are:

- We wanted to allow `TreeItem` to be any shape. Just taking a few properties off the `TreeItem` for rendering feels bad.
- Separating concerns of storing data(`T`, extension will do that) and rendering (all logic in Provider).
- From extension author's POV, extending `TreeItem` and then upcasting the `TreeItem` to their own class in `resolveChildren` and the registered commands callback feels weird.
- The extension looks pretty good with TS2's [tagged union](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#tagged-union-types): https://github.com/octref/explorer-api-example/blob/master/dep/src/extension.ts

---

> The node provider should have a way to signal to the main side that a node/substree has changes

Agreed. I also wanted this. It's not there as Kai was initially averse to it, but he now seems to support it. We have a refresh button on top of the viewlet. 

---

> Tree elements should support icons - in one or the other way

Decided to leave that for next iteration. Wanted to get some feedback from users before deciding how to do it.

---

> We should consider adding something like paging to the resolveChildren-call. I see this API being used for large sets or sets that are slow to resolve.

We are showing the progress bar if the `resolveChildren` is taking long, similar to what the search viewlet is doing. I feel we can delegate paging to extension authors. Resolve first page -> notify VSCode of change in the subtree when they resolve next page.

Better to wait for some feedback before adding it, I think.

---

> The API lacks a way to reveal, select, focus a node in a tree.

Yep. But while I was writing my 3 examples I never found the need for it. And I can't think of a use case for DB Explorer, which we are designing this for.

So better wait for next iteration when we understand the use cases more.

---

Will sync with Kai/Daniel this afternoon and talk through your concerns.

jrieken replied:
You don't need to up-cast anything when your define it as `<T extends TreeItem>`. It makes an easy to understand API and *most importantly* it aligns with other parts of the API. Keeping things consistent (look at quick pick) is key to make an easily understandable API. Your sample proofs that the *label provider* indirection is artificial as the actual data is stored in the [node](https://github.com/octref/explorer-api-example/blob/master/dep/src/extension.ts#L34) and not computed by a provider.

Tyriar replied:
We actually had it earlier as you're requesting and @kieferrm convinced us to go in this direction.

> Your sample proofs that the label provider indirection is artificial as the actual data is stored in the node and not computed by a provider

There are a few benefits to doing it this way:

- Instead of passing in some TreeItem data structure, the functions work on the actual backing object itself. This removes the need to go and construct a bunch of node objects but instead derive the label from the object you're working on. This extends to the click command, has children, etc. 
- In the previous paradigm the TreeItem typically required some `data` attribute to disambiguous duplicate labels that appear at multiple places within the tree. Consider a class explorer that have multiple nested objects that may have the same labels but has different click, children etc, 
  
  ```
  class foo
    int a
    class foo
      int b
  ```

  When it comes to provide the second `foo`'s nodes, you need some additional context to handle this. This came up in the examples @octref wrote, so it would be typical to add a `TreeItem.data` property anyway, this just cuts the middle man.
- It becomes very easy to switch all labels to use a different function by simply changing the function and triggering an update for example. 

There's some more context on this in https://github.com/octref/vscode/issues/4 and https://github.com/octref/vscode/issues/5.

I was also originally opposed to this way but it's definitely grown on me and I think it's quite a nice way to work on a tree structure. I agree consistency very important - just some background/insights into the current situation.

jrieken replied:
Having worked on Eclipse before  I am very well aware of how the [LabelProvider](http://help.eclipse.org/kepler/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjface%2Fviewers%2FLabelProvider.html) and [ContentProvider](http://help.eclipse.org/kepler/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjface%2Fviewers%2FIContentProvider.html) patterns work. The question I am asking and not getting feedback on is how does it align with QuickPick, how would a tree'ish quick pick data structure look like, and how does it generally fit into the API?

octref replied:
@jrieken 

> how does it align with QuickPick

I admit it is very different from QuickPick API. I think it's more aligned to other Providers like `CodeLensProvider`, or `DocumentSymbolProvider`.

> how would a tree'ish quick pick data structure look like

That's what I like about the current design. It doesn't care what your TreeItem data-structure look like. As long as your provider can `getLabel` for your TreeItem, you can work with our API.

> how does it generally fit into the API?

It doesn't resemble any existing API. Previously we did have a version like this:

```ts
export interface TreeExplorerNodeProvider {
    provideRootNode(): Thenable<TreeExplorerNode>;
    resolveChildren(node: TreeExplorerNode): Thenable<TreeExplorerNode[]>;
}

export interface TreeExplorerNode {
    label: string;
    shouldInitiallyExpand: boolean;
    onClickCommand: string;
}
```

but we switched to the `TreeExplorerNode`-less design for reasons @Tyriar mentioned.

octref commented on src/vs/workbench/api/node/extHost.protocol.ts:
```diff
@@ -257,6 +264,12 @@ export abstract class ExtHostEditorsShape {
 	$acceptTextEditorRemove(id: string): void { throw ni(); }
 }
 
+export abstract class ExtHostTreeExplorersShape {
```
`InternalTreeExplorerNode` doesn't have any method. It's the way we communicate between `ExtHost/MainThread`. Node will be provided and stored at `ExtHost` side. For each node we use `getLabel` etc to get info from the node, and construct a slim internal node to send to `MainThread`.

octref replied:
Oops sorry. Actually I designed with this in mind and should use `InternalTreeExplorerNodeContent`, which is an interface.

octref commented on src/vs/platform/extensionManagement/common/extensionManagement.ts:
```diff
@@ -85,6 +91,7 @@ export interface IExtensionContributions {
 	menus?: { [context: string]: IMenu[] };
 	snippets?: ISnippet[];
 	themes?: ITheme[];
+	explorer?: ITreeExplorer;
```
Decided to start with this, because I don't know which path we'll take if we were to allow more complex explorer contribution in the future.

1. Allow multiple sections (tree or non-tree) like debug

  ![image](https://cloud.githubusercontent.com/assets/4033249/19822810/3c78dcd8-9d1a-11e6-8061-4d029cf16499.png)

2. Allow multiple explorers per extenseion

Better to keep it minimal for first iteration I think.

jrieken replied:
[needs thinking] The problem is that this means no extension can ever add more than one tree-viewlet

octref replied:
Even if we change this part later, it would be an easy update from extension side. I found use cases for multiple viewlets less unlikely than a single multi-section viewlet.

jrieken commented:
Thinking more about this, have we considered such an API idea? Wwe could do something more similar to quick pick. I am thinking of a data model that can be shown in both quick pick and/or in an explorer-style. That would mean, we have a type ala

```ts
interface Item {
  label: string;
  command: Command;
  resolveChilren(): Promise<TreeItem[]> | TreeItem[];
}
```

which you can use in `showQuickPick` and in a method that populates a tree. I am bringing this up because it has been mentioned more than once that folks want to show a tree in quick pick and I would hate completely different APIs for these very similar problems.

jrieken commented:
@octref Can you go ahead and resolve merge conflicts and move the API additions from `vscode.d.ts` to `vscode.proposed.d.ts` (added with c510d53da79f7184038378f511b7459321f5b217). Plan is to make this *proposed api* first, merge quickly, and iterate further on this.

*Note* on proposed API: The story is that proposed API is defined in a separate file and extension writers have to (manually) pull that file into their workspace. Also, in `package.json` you must opt in to this (`enableProposedApi: true`). Therefore, make sure to only expose proposal API feature when enabled, like done with the [`sampleFunction`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/node/extHost.api.impl.ts#L277)

octref commented:
Moved the API to proposed dts file.

Originally I didn't have a strong preference as to which way we take, but now looking at the API again from a high level, I do lean toward this approach:

```ts
export interface TreeExplorerNodeProvider {
	provideRootNode(): TreeExplorerNode | Thenable<TreeExplorerNode>;
	resolveChildren(node: TreeExplorerNode): TreeExplorerNode[] | Thenable<TreeExplorerNode[]>;
}

export interface TreeExplorerNode {
	label: string;
	hasChildren?: boolean;
	onClickCommand?: Command;
}
```

In the extension, doing this:

```ts
// Provider
getLabel(node: DepNode): string {
  switch (node.kind) {
    case 'root':
      return '';
    case 'node':
      return node.moduleName;
    case 'leaf':
      return node.moduleName;
  }
}
```

or this:

```ts
class Root implements TreeExplorerNode {
  label: '';
}

class Node implements TreeExplorerNode {
  get label() { return this.moduleName; }
}

class Leaf {
  get label() { return this.moduleName; }
}
```

is more a matter of preference.

But I do observe a strong consistency in the API about the `registerSomethingProvider`, and then `Provider.provideSomething(): Something | Thenable<Something>`

```ts
function registerCompletionItemProvider(selector: DocumentSelector, provider: CompletionItemProvider, ...triggerCharacters: string[]): Disposable;
function registerCodeActionsProvider(selector: DocumentSelector, provider: CodeActionProvider): Disposable;
function registerCodeLensProvider(selector: DocumentSelector, provider: CodeLensProvider): Disposable;

interface CompletionItemProvider {
  provideCompletionItems(document: TextDocument, position: Position, token: CancellationToken): CompletionItem[] | Thenable<CompletionItem[]> | CompletionList | Thenable<CompletionList>;
  resolveCompletionItem?(item: CompletionItem, token: CancellationToken): CompletionItem | Thenable<CompletionItem>;
}
interface CodeActionProvider {
  provideCodeActions(document: TextDocument, range: Range, context: CodeActionContext, token: CancellationToken): Command[] | Thenable<Command[]>;
}
interface CodeLensProvider {
  provideCodeLenses(document: TextDocument, token: CancellationToken): CodeLens[] | Thenable<CodeLens[]>;
  resolveCodeLens?(codeLens: CodeLens, token: CancellationToken): CodeLens | Thenable<CodeLens>;
}
```

and I'd rather not break that consistency.

I pushed an example using the old `TreeExplorerNode` API: https://github.com/octref/explorer-api-example/blob/master/dep2/src/extension.ts
Example using current approach API: https://github.com/octref/explorer-api-example/blob/master/dep/src/extension.ts

Looking for some feedback.
CC @kieferrm

jrieken commented on src/vs/workbench/services/activity/common/activityService.ts:
```diff
@@ -70,4 +70,24 @@ export interface IActivityService {
+	 * Get the external viewlet id that is about to open.
+	 */
+	getExternalViewletIdToOpen(): string;
+}
```
[needs thinking] @bpasero will know better, but I have a strong feeling that the `IActivityService` isn't the right place for viewlet specifics. Isn't there an `IViewletService` and wouldn't that be the better place for this?

octref replied:
To my understanding `SideBarPart/IViewletService` is more low-level and just does composite construction, whereas `ActivityBarPart/IActivityService` knows all viewlets and which ones are enabled. So I feel the logic should be here instead of in `IViewletService`, but the name `IActivityService` is indeed confusing.

jrieken commented on src/vs/workbench/workbench.main.ts:
```diff
@@ -19,6 +19,9 @@ import 'vs/editor/browser/editor.all';
 import 'vs/platform/actions/browser/menusExtensionPoint';
 
+// External Explorers
+import "vs/platform/explorers/browser/explorerExtensionPoint";
```
[needs change] Please move this file into the `workbench` folder, only shared services (between editor/workbench) and things around it should be in `platform`

jrieken commented on src/vs/workbench/browser/viewlet.ts:
```diff
@@ -173,6 +177,16 @@ export class ViewletRegistry extends CompositeRegistry<Viewlet> {
 	/**
+	 * Registers multiple external viewlets.
+	 */
+	public registerExternalViewlets(descriptors: ViewletDescriptor[]): void {
```
@bpasero for more feedback on this

bpasero replied:
Yes, I also found this weird when initially scanning the PR and suggest to fold this into the same solution we already have. There is no need to distinguish external and "internal" viewlets if possible.

octref replied:
Removed this and made the loading of external viewlets this way:

- Load stocked ones and display icons on `ActivityBarPart`
- `ActivityBarPart` add in external viewlets' icons once all extensions are loaded

jrieken commented:
@octref I have added two kind of feedback `[needs change]` and `[needs thinking]`. The address the former unless you have strong objections/reasons. The latter we might tackle *post merge*.

octref commented:
Addressed all comments. One question I have: https://github.com/Microsoft/vscode/pull/14048/files#r86076291

jrieken commented:
@octref The picker to select and actually show a custom viewlet, is that the current story for configuring the activity bar or why do we have that? I took me a while to understand why my viewlet didn't show up. Also, when the extension is uninstalled, there were some exceptions when reading from the memento.

octref commented:
@jrieken Yes that was the story. Maybe I can let a new extension show up by default to reduce confusion.

Let me look into uninstall.

octref commented:
@jrieken Which error did you see specifically? I tried uninstall and restart, didn't see anything.

bpasero commented:
@octref expect more feedback from me tomorrow, today was swamped with meetings, sorry.

octref commented:
@bpasero Sure, no worries.

bpasero commented on src/vs/workbench/parts/explorers/browser/treeExplorerActions.contribution.ts:
```diff
@@ -0,0 +1,57 @@
+/*---------------------------------------------------------------------------------------------
```
@octref not sure if you care about it but this file could move into the `common` layer since it does not seem to depend on anything browser related.

bpasero commented on src/vs/workbench/parts/explorers/media/treeExplorer.contribution.css:
```diff
@@ -0,0 +1,18 @@
+/*---------------------------------------------------------------------------------------------
```
@octref please check for typical CSS we use in trees. For example the items currently are not looking centered because they miss a `line-height: 22px` instruction. 

![image](https://cloud.githubusercontent.com/assets/900690/19959225/6484bf0a-a1a5-11e6-89c6-e991c6cf9c7f.png)

bpasero commented on src/vs/workbench/browser/explorerExtensionPoint.ts:
```diff
@@ -0,0 +1,60 @@
+
+		const getIconRule = (iconPath) => { return `background-image: url('${iconPath}')`; };
+		if (icon) {
+			const iconClass = `.monaco-workbench > .activitybar .monaco-action-bar .action-label.${treeExplorerNodeProviderId}`;
```
@octref shouldn't the icon rule be created on `VIEWLET_ID_ROOT + treeExplorerNodeProviderId`? How do we otherwise prevent someone from overwriting the icon for the explorer?

Also you have to be carful when taking a user defined string as CSS class. Maybe ping @aeschli on his use of CSS escaping (see https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/themes/electron-browser/themeService.ts#L603)

aeschli replied:
Yes, @bpasero is right, some characters are not valid in a css selector. Check out https://mathiasbynens.be/notes/css-escapes.
As you are in the Electron renderer process you can use window['CSS'].escape(str). We might want to move that to a util at some point

octref replied:
Decided to limit `viewletId` to alphanumeric with `-` and `_`, and show error if viewlet has invalid id.

bpasero commented on src/vs/workbench/browser/viewlet.ts:
```diff
@@ -153,7 +153,7 @@ export abstract class ViewerViewlet extends Viewlet {
 export class ViewletDescriptor extends CompositeDescriptor<Viewlet> {
 
-	constructor(moduleId: string, ctorName: string, id: string, name: string, cssClass?: string, order?: number) {
+	constructor(moduleId: string, ctorName: string, id: string, name: string, cssClass?: string, order?: number, public isExternal: boolean = false) {
```
@octref `isExternal` is a bit confusing, can we maybe call this `isExtension`?

bpasero commented on src/vs/workbench/browser/explorerExtensionPoint.ts:
```diff
@@ -0,0 +1,60 @@
+			VIEWLET_ID_ROOT + treeExplorerNodeProviderId,
+			treeLabel,
+			treeExplorerNodeProviderId,
+			-1, // External viewlets are ordered by enabling sequence, so order here doesn't matter.
```
@octref `Extension viewlets...`

bpasero commented on src/vs/workbench/parts/explorers/common/treeExplorer.ts:
```diff
@@ -0,0 +1,7 @@
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+export const VIEWLET_ID_ROOT = 'workbench.view.customTreeExplorer.';
```
@octref since you seem to use this always as a prefix, might be more elegant to just expose a function `toViewletId()` that does the concatenation?

octref replied:
Good suggestion.

bpasero commented on src/vs/workbench/parts/explorers/browser/treeExplorerActions.contribution.ts:
```diff
@@ -0,0 +1,57 @@
+			const { isEnabled, treeLabel } = infoForExternalViewlets[viewletId];
+			picks.push({
+				id: viewletId,
+				label: (isEnabled ? 'Disable ' : 'Enable ') + treeLabel
```
@octref you are messing with nls here!

bpasero commented on src/vs/workbench/parts/explorers/browser/treeExplorerActions.contribution.ts:
```diff
@@ -0,0 +1,57 @@
+		}
+
+		return TPromise.timeout(50 /* quick open is sensitive to being opened so soon after another */).then(() => {
+			this.quickOpenService.pick(picks, { placeHolder: 'Select Viewlet to toggle', autoFocus: 2 }).then(pick => {
```
@octref multiple things:
* nls!
* I suggest to not call it Viewlet but rather similar to the action that it was opened from ("Custom Explorer")
* please use the enum and not the number it represents for `autoFocus`

bpasero commented on src/vs/workbench/services/activity/common/activityService.ts:
```diff
@@ -70,4 +71,19 @@ export interface IActivityService {
 	clearActivity(compositeId: string): void;
-}
\ No newline at end of file
+
```
@octref this Viewlet related code should move into `IViewletService` or if that is impossible for some reason maybe into a new `IExtensionViewletService`

bpasero commented on src/vs/workbench/parts/explorers/browser/treeExplorerViewletService.ts:
```diff
@@ -0,0 +1,76 @@
+/*---------------------------------------------------------------------------------------------
```
@octref I find the naming of this service weird and hard to understand what it does. can we maybe rename it to something that better tells us what the purpose is? I see the name is `ITreeExplorerViewletService` but it gets contributed as `customViewletService` which seems weird. Mayb `IExtensionTreeService`? Afaik it seems only tree related but not viewlet related.

bpasero commented on src/vs/workbench/parts/explorers/browser/treeExplorerViewlet.ts:
```diff
@@ -0,0 +1,88 @@
+		@IInstantiationService private instantiationService: IInstantiationService,
+		@IActivityService private activityService: IActivityService
+	) {
+		super(VIEWLET_ID_ROOT + TreeExplorerViewlet._idCounter, telemetryService);
```
@octref this is very confusing and fragile. the viewlet ID is now suddenly something dynamic and not static? and then you get an external ID from somewhere and return it from `getId()`? All of this has to change :-/

bpasero commented on src/vs/workbench/electron-browser/workbench.ts:
```diff
@@ -228,9 +228,14 @@ export class Workbench implements IPartService {
-				const viewletTimerEvent = timer.start(timer.Topic.STARTUP, strings.format('Opening Viewlet: {0}', viewletId));
-				compositeAndEditorPromises.push(this.sidebarPart.openViewlet(viewletId, false).then(() => viewletTimerEvent.stop()));
+			// If external viewlet is the last active viewlet, defer its construction until all extensions are loaded
+			if (!viewletRegistry.getViewlet(viewletId)) {
```
@octref this is ugly and should be removed. instead the viewlet service should be fit enough to handle both extension viewlets and normal viewlets. if you clean that up you will not need this if-else condition at all 👍

bpasero commented on src/vs/workbench/browser/parts/activitybar/activitybarPart.ts:
```diff
@@ -8,6 +8,7 @@
 import 'vs/css!./media/activitybarpart';
 import nls = require('vs/nls');
 import { TPromise } from 'vs/base/common/winjs.base';
+import Event, { Emitter } from 'vs/base/common/event';
```
@octref I expect some code from here to move out to `IViewletService` or `IExtensionViewletService`. it feels lots of things that should not be activity bar related are now inside this part. To me, the part should not really distinguish between normal viewlets and extension viewlets, but just do its business of rendering a set of viewlets and their activity.

bpasero commented:
@octref submitted a first set of feedback. overall my biggest issue is that lots of responsibility is pushed to activity bar and activity service that should rather live in `IViewletService` or maybe a new `IExtensionViewletService`. 

* [ ] Adding an external viewlet or removing it removes active badges on the activity bar
* [ ] Disabling the external viewlet should jump back to the default viewlet (explorer)
* [ ] Disabling an external viewlet and reloading opens the workbench with no viewlet at all
* [ ] Enabling an external viewlet should open it
* [ ] External viewlet loading should hook into the progress bar of the sidebar so that you see the viewlet is actually doing something
* [ ] An external viewlet should be able to show content to the activity badges
* [ ] Did you try having multiple external viewlets from a single extension or multiple extensions in one window?

jrieken commented:
Re https://github.com/Microsoft/vscode/pull/14048#issuecomment-258089341 - Can we agree on only addressing issues that a problematic at this stage and not implement additional requirements (like badges or multiple viewlets etc)

bpasero commented:
Yeah, consider my feedback in the code review a "MUST fix before merge" and the items listed in the comment as bugs that can be fixed after merging.

octref commented:
@bpasero I looked through the feedback. Seems reasonable and will address today, I have a few questions though:

- Currently `SidebarPart` is `IViewletService`, this is a bit confusing to me. Do you mind if I rename it and make a real `IViewletService` that handles viewlet specifics?
- > the viewlet ID is now suddenly something dynamic and not static? and then you get an external ID from somewhere and return it from getId()? All of this has to change :-/

  That was one of my question: https://github.com/Microsoft/vscode/pull/14048/files#r86076291. `TreeExplorerViewlet` doesn't know its ID until extension loading happens. How can I make it static?

octref commented:
And

> Did you try having multiple external viewlets from a single extension or multiple extensions in one window?

I tested with multi-extensions. As for multi-explorers in one extension, see this: https://github.com/Microsoft/vscode/pull/14048#discussion_r84432594

bpasero commented:
@octref I am fine if you create a new class `ViewletService` that implements `IViewletService` and implements some of the methods using `SidebarPart`. We actually do the very same with `IEditorService` and `EditorPart`, so it is a good pattern to follow.

As for the ID I see the issue and it might require changes in how viewlets are created. As a first step, can we not pass over the ID from the viewlet descriptor to the viewlet so that you do not have to do this ugly counter thing? My understanding is that the viewlet descriptor gets registered when the extensions are loaded so that guy knows the right ID to use and can pass it over?

octref commented:
@bpasero 

Addressed all comments, and some of your additional feedbacks. Will do progress bar tomorrow and leave badge for post merge.

Let me know what else you want to me to change.

bpasero commented:
@octref looks a lot better, there is only one thing I am not sure about in workbench.ts you seem to have changed the logic about how to restore a viewlet.

Previously: the check for `shouldRestoreSidebar` is used to find out if the default viewlet (explorer) should be opened or actually the last viewlet that was open. `shouldRestoreSidebar` will be true when running out of sources (to speed up development) or when a refresh of the window was initiated (e.g. after installing an extension).

Now: you seem to never open a viewlet if `shouldRestoreSidebar` is false. That would mean anytime I open Code I would see the sidebar collapsed. That would be pretty bad and a bug. Can we get back to the previous behaviour?

Also I wonder if we can move the logic about extension viewlets into the viewlet service. `IViewletService.openViewlet()` should be the one that checks if a viewlet is external and defer its opening until it is loaded. Otherwise everyone that wants to open a viewlet has to be aware of the fact that the viewlet can be external and not yet loaded.

octref commented:
If I understand correctly:

- if `shouldRestoreSidebar` is true
  - restore last opened viewlet if possible
  - restore file explorer in cases like last opened viewlet is disabled/uninstalled
- if `shouldRestoreSidebar` is false
  - restore file explorer

is this what you meant?

> move the logic about extension viewlets into the viewlet service

Sure, will do.

bpasero commented:
@octref I realize that the name `shouldRestoreSidebar` is confusing so feel free to change it. Yes, the only time we do NOT open a viewlet is when the sidebar is hidden. Otherwise we restore the explorer if `shouldRestoreSidebar` is false and the last used viewlet otherwise.

octref commented:
@bpasero Made changes as you requested. Decided to put the logic in a dedicated `restoreViewlet`, as the logic is heavy & need to call registry multiple times. `openViewlet` should be used after editor startup when all extensions are loaded.

LiPengfei19820619 commented:
Good News. Thank you.
